### ISSUE TITLE
[Feature] Add on transactions window the transactions received too

### ISSUE THREAD

**Issue by bitcoinuser** [getAlby/lightning-browser-extension#2201]
### Feature description

Add on transactions window the transactions received too. For now only appear the transactions we sent.

**Comment by escapedcat** [comment]
@reneaaron @rolznz wdyt?

**Comment by reneaaron** [comment]
@lujakob Yes, this is a feature we are currently planning. Happy to get your help with that! 🙌

As we are currently thinking about unifying the incoming / outgoing transactions I think this page could be a good fit to implement this as a proof of concept before changing the Home screen to the unified view as well. 

For the unified transaction lists we would need to introduce some "+" and "-" icons to recognize incoming and outgoing transactions respectively. 

Also we could think of filtering the outgoing transactions on an account basis as the account id is now also saved for outgoing payments. 

To summarize: 
 - Load the incoming transactions on the transactions page, too
 - Filter the outgoing transactions per account
 - Some minor UI changes to make it obvious which transactions are incoming / outgoing
 
Does that sound like a plan to you @lujakob?

**Comment by lujakob** [comment]
@reneaaron Sounds good. 

The list of transactions "unified" means there is only one list showing all transactions. To separate outgoing/incoming we would add +/- icons. (I remember someone talking about tabs for incoming/outgoing. This idea is discarded I guess?)

Can I already get started on this, or we have dependencies or waiting for further design details?

**Comment by lujakob** [comment]
@reneaaron @escapedcat  I checked the code. Some thoughts on the approach of unifying transactions, as it is relevant to pagination. To unify incoming and outgoing transactions I have to get entities from two sources. Outgoing transactions are IndexDB transactions and incoming transactions are invoices from the connector. The connector invoices are unpaginated and come altogether. The IndexDB payments could be paginated, however I think it could be tricky to integrate the two sources and have a feasible paginated source. The easy way would be to get all payments, all invoices, combine and sort them, and send only the relevant "page" as a response to the component. However this would render the IndexDB performance improvement irrelevant. Only the UX improvement of a pagination for the user would be left. (Showing e.g. 100 transactions at a time) I guess as the data from IndexDB doesn't come from a slow network request, but from a local memory source, it could still be fast enough. If not, getting reliable paginated data from two sources could get pretty tricky. What do you think?

**Comment by lujakob** [comment]
@reneaaron @escapedcat At which part of the code should the integration/aggregation of the two sources happen? I would suggest to create a folder `extension/background-script/actions/transactions/list.ts`. Even though the transactions are an aggregated source of entities from `extension/background-script/actions/ln/invoices.ts` and `extension/background-script/actions/payments/all.ts` I think it makes sense, because from a frontend/component point of view I want to request the data in the same way as I request any other entity source using the api interface. At least I would not aggregate/integrate the data in the component, as this might be reused. What do you think?

**Comment by reneaaron** [comment]
> The list of transactions "unified" means there is only one list showing all transactions. To separate outgoing/incoming we would add +/- icons. (I remember someone talking about tabs for incoming/outgoing. This idea is discarded I guess?)

Yes. The bigger plan is to unify them. We'll have to see if that works out as there are some problems that need to be solved for that. 

> Can I already get started on this, or we have dependencies or waiting for further design details?

I don't think there are many new things in here (except for the icons) so I think we can go ahead. 

### Pagination

After some more thoughts that could really be a tricky problem. We would need to fetch and fill the unified array from the two sources until we have enough transactions to return. 🤔 I also think that this should be done in a central place, there should be one method that allows you get unified transaction data that is sorted & can be paginated.  

> However this would render the IndexDB performance improvement irrelevant. 

Which improvement do you mean here? 

> I think it makes sense, because from a frontend/component point of view I want to request the data in the same way as I request any other entity source using the api interface. At least I would not aggregate/integrate the data in the component, as this might be reused. What do you think?

+1 for creating some new apis and not doing that in the component. The Home view will hopefully also be able to use this data soon.

**Comment by lujakob** [comment]
> > However this would render the IndexDB performance improvement irrelevant.
> 
> Which improvement do you mean here?

One reason in favour of paginated data is it improves time loading data. In normal SPA it improves network request time if data is fetched paginated. In our case we don't have network requests, but query the IndexDB. I cannot paginate the IndexDB payments however, because I need to merge the results with invoices. If a page has 100 transactions, and sorted by `createdAt` it will consist of 87 payments from IndexDB and 13 invoices from the connector, this ratio can only be known after merging and sorting the data. So the information for querying paginated data (pageSize, currentPage) will vary for each page and is very hard to keep track of and calculate for each new page. For this reason we have to fetch all payments, merge with invoices and return a paginated result set. You know what I mean? Any opinions on this?

**Comment by reneaaron** [comment]
I was thinking if we could load the 2 sources incrementally:

So while we don't have enough results (`pagesize`) we would could load from the 2 sources incrementally and don't waste any requests (to the connector as well as the indexdb). See this very incomplete example: 😅

```js
let incomingPage, outgoingPage = 0;

while(results.length < pagesize) {
  const incoming = loadIncoming(incomingPage);
  const outgoing = loadOutgoing(outgoingPage);
  
  // TODO loop through the results and see which items need to be added to the result
  // Remember the last date of each collection to avoid fetching the same page multiple times (maybe that can also be solved with Swr?)
}
```

That way we wouldn't load more data than absolutely needed and the code could be wrapped in some `getTransactions()` method that hides this complexity from the consumer completely. Does that make sense?

**Comment by lujakob** [comment]
@reneaaron thanks for your proposals, but your suggestion does't help me. But it got me knew ideas, and new problems.

First of all, I did some further research and a lot of connectors don't provide `getInvoices` functionality and show an error instead. I checked the docs for Lightning Core and they do not provide pagination yet. I created an [issue](https://github.com/ElementsProject/lightning/issues/6098#issuecomment-1471529535) and apparently there is a workaround, but not sure if it's possible. Another problem is, the enable a proper pagination component, the total amount of invoices is necessary. Afaik, this is not provided. I think for Core Lightning connector we're not able to provide it yet.

Regarding LND, it is possible to receive paginated data, however it is not possible to get the total number of invoices. Without, I am able to have a "load more" style pagination, but it's not possible to know and display the total number of pages.

Second: even though we might have all data and params to paginate invoices, still we have the issue of combining the data. And from my understanding of it there is a problem I cannot overcome (yet). I am talking about the `backend-script/actions` part here. 

The first challenge is, the ratio of numbers of payments and invoices is not predictable for each page. For the first page this is irrelevant. Assuming a pageSize of 100, I just request 100 payments and 100 invoices, join the results, order by date and pick the first 100. Then I know I have e.g. 70 payments and 30 invoices in it. So in the response to the UI I send the first 100 transactions and pass along the offsets separately. The invoicesOffset is 30 and the paymentsOffset is 70. For page 2 I pass those in the request and I can just do the same thing again, getting 100 each starting with the relevant offset. So for the second result set of 200 (100 payments and 100 invoices) I sort by date and pick the first 100 and find 60 payments and 40 invoices. I add those to the offset from before to have a new invoiceOffset of 70 (30 + 40) and a paymentsOffset of 130 (70 + 60).

So this is no problem if I go from one page to the next. The problem however is, if I want to jump from page 1 to any other page than the next. So if I jump from page 1 to let's say page 4, I cannot now what each offset is, because they are not predictable. I can only build them up one after the other. Usually in a pagination of a single source I can tell the offset for page 4 with pageSize 100 is 4 * 100 = 400. But with two varying sources I cannot predict. Also navigating back would be a problem I guess. 

To make it work I would need to combine the whole transactions data (payments and invoices) sort them, and from this source get the paginated data. 

So we could for example on page load or connector change get the first page and in the background iterate over all invoices requests. Once finished combine them with payments and hold this in the Zustand state. Once this is done, show the paginator. This is already quite complex. I already don't really like it. But then what happens on account change? The connector changes, all invoices get loaded again. This has some delays. I'm not really happy with this.

So I don't see a good solution. What do you guys think?
